### PR TITLE
[FLINK-25230][table-planner] Replace RelDataType with LogicalType serialization

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This is another step of hardening the JSON representation. Instead of exposing Calcite's type stack, we rely on Flink `LogicalType` instead. This ensures that only one type system needs to be maintained in the future. It will also make Calcite upgrades easier.

Some tests were adapted to represent what is actually needed during a plan restore.

## Brief change log

- Introduce `LogicalType` <-> `RelDataType` converter for the future
- Serialize `LogicalType`

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/typeutils/LogicalRelDataTypeConverter.java [hunk 1]
```diff
@@ -0,0 +1,649 @@
...
+            case TINYINT:
+                return new TinyIntType(false);
+            case SMALLINT:
+                return new SmallIntType(false);
+            case INTEGER:
+                return new IntType(false);
+            case BIGINT:
+                return new BigIntType(false);
+            case DECIMAL:
+                if (relDataType.getScale() < 0) {
```
Comment (matriv) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/typeutils/LogicalRelDataTypeConverter.java:483]: What's the case that this code addresses? Forgive me if I didn't see, is it also tested?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/JsonSerdeMocks.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    static <T> T toObject(SerdeContext serdeContext, String json, Class<T> clazz) {
+        final ObjectReader objectReader = JsonSerdeUtil.createObjectReader(serdeContext);
+        try {
+            return objectReader.readValue(json, clazz);
+        } catch (IOException e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    private JsonSerdeMocks() {
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/JsonSerdeMocks.java:73]: nit: maybe move to the top of the class.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/JsonSerdeMocks.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.flink.table.utils.CatalogManagerMocks;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.core.JsonProcessingException;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectReader;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectWriter;
+
+import java.io.IOException;
+
+/** Mocks and utilities for serde tests. */
+class JsonSerdeMocks {
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/JsonSerdeMocks.java:36]: could also be `final` as a utility class?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java [hunk 4]
```diff
@@ -132,101 +130,35 @@
...
-                                BigDecimal.valueOf(3),
-                                new SqlIntervalQualifier(
-                                        TimeUnit.SECOND, TimeUnit.SECOND, SqlParserPos.ZERO)),
+                                        TimeUnit.SECOND, 2, TimeUnit.SECOND, 6, SqlParserPos.ZERO)),
                         rexBuilder.makeDateLiteral(DateString.fromDaysSinceEpoch(10)),
                         rexBuilder.makeDateLiteral(new DateString("2000-12-12")),
                         rexBuilder.makeTimeLiteral(TimeString.fromMillisOfDay(1234), 3),
                         rexBuilder.makeTimeLiteral(TimeString.fromMillisOfDay(123456), 6),
                         rexBuilder.makeTimeLiteral(new TimeString("01:01:01.000000001"), 9),
-                        rexBuilder.makeTimeWithLocalTimeZoneLiteral(
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java:228]: Why did you remove this, and also before some interval related ones?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RelDataTypeJsonSerdeTest.java [hunk 5]
```diff
@@ -46,83 +35,113 @@
...
+        final List<RelDataType> types =
                 Arrays.asList(
                         FACTORY.createSqlType(SqlTypeName.BOOLEAN),
                         FACTORY.createSqlType(SqlTypeName.TINYINT),
                         FACTORY.createSqlType(SqlTypeName.SMALLINT),
                         FACTORY.createSqlType(SqlTypeName.INTEGER),
                         FACTORY.createSqlType(SqlTypeName.BIGINT),
-                        FACTORY.createSqlType(SqlTypeName.DECIMAL, 3, 10),
-                        FACTORY.createSqlType(SqlTypeName.DECIMAL, 0, 19),
-                        FACTORY.createSqlType(SqlTypeName.DECIMAL, -1, 19),
```
Comment (matriv) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RelDataTypeJsonSerdeTest.java:78]: Shouldn't we test the negative value?

**Review by matriv — COMMENTED** [review]
Thank you @twalthr Left a few comments/questions.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java (hunk 4 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java:228]: We don't support time with timezone. I also wanted to simplify the tests a bit. The tests were creating RexLiterals that Flink never generates.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RelDataTypeJsonSerdeTest.java (hunk 5 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RelDataTypeJsonSerdeTest.java:78]: correct me if I'm wrong, but why should we support negative precision or scale that is smaller than precision? it looks like a bug in the tests to me.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java (hunk 4 above)
Comment (matriv) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java:228]: It's LTZ, that's why I asked.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RelDataTypeJsonSerdeTest.java (hunk 5 above)
Comment (matriv) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RelDataTypeJsonSerdeTest.java:78]: I think so, that's why I asked here: https://github.com/apache/flink/pull/18342#discussion_r783771529

Comment (twalthr) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RelDataTypeJsonSerdeTest.java:78]: This covers negative scale. Negative scale we can support. But what should negative precision mean?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/typeutils/LogicalRelDataTypeConverter.java (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/typeutils/LogicalRelDataTypeConverter.java:483]: True, I added a test.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java (hunk 4 above)
Comment (twalthr) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RexNodeSerdeTest.java:228]: but time and not timestamp. I don't think that we will ever support time LTZ.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RelDataTypeJsonSerdeTest.java (hunk 5 above)
Comment (matriv) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/serde/RelDataTypeJsonSerdeTest.java:78]: I can't think of such a case, doesn't make sense.

**Review by matriv — APPROVED** [review]
LGTM Thanks @twalthr!
